bundler/lib/bundler/compact_index_client/cache.rb

```diff
@@ -7,16 +7,46 @@ class CompactIndexClient
     class Cache
       attr_reader :directory
 
-      def initialize(directory)
+      def initialize(directory, updater)
         @directory = Pathname.new(directory).expand_path
-        info_roots.each {|dir| mkdir(dir) }
-        mkdir(info_etag_root)
+        @updater = updater
+
+        @info_root = mkdir("info")
+        @special_characters_info_root = mkdir("info-special-characters")
+        @info_etag_root = mkdir("info-etags")
       end
 
       def names
-        lines(names_path)
+        fetch("names", names_path, names_etag_path)
       end
 
+      def versions
+        fetch("versions", versions_path, versions_etag_path)
+      end
+
+      def info(name, remote_checksum)
+        unless remote_checksum
+          Bundler::CompactIndexClient.debug { "skipping updating info for #{name} since it is missing from versions index" }
+          return read(info_path(name))
+        end
+
+        path = info_path(name)
+        local_checksum = SharedHelpers.checksum_for_file(path, :MD5)
+        if local_checksum == remote_checksum
+          Bundler::CompactIndexClient.debug { "skipping updating info for #{name} since the versions checksum matches the local checksum" }
+          read(path)
+        else
+          Bundler::CompactIndexClient.debug { "updating info for #{name} since the versions checksum #{remote_checksum} != the local checksum #{local_checksum}" }
+          fetch("info/#{name}", path, info_etag_path(name))
+        end
+      end
+
+      def available?
+        versions_path.file? && versions_path.size.nonzero?
```

Comment:
Probably needs to try fetching to be useful, but currently the parser doesn't call this anyway.